app/models/permitted_params.rb

```diff
@@ -262,6 +236,155 @@ def wiki_content
+  end
+
+  def version
+    # `version_settings_attributes` is from a plugin. Unfortunately as it stands
+    # now it is less work to do it this way than have the plugin override this
+    # method. We hopefully will change this in the future.
+    params.require(:version).permit(:name,
+                                    :description,
+                                    :effective_date,
+                                    :due_date,
+                                    :start_date,
+                                    :wiki_page_title,
+                                    :status,
+                                    :sharing,
+                                    :custom_field_value,
+                                    version_settings_attributes: [:id, :display, :project])
+  end
+
+  def comment
+    params.require(:comment).permit(:commented, :author, :comments)
+  end
+
+  # `params.fetch` and not `require` because the update controller action associated
+  # with this is doing multiple things, therefore not requiring a message hash
+  # all the time.
+  def message(instance = nil)
+    if instance && current_user.allowed_to?(:edit_messages, instance.project)
+      params.fetch(:message, {}).permit(:subject, :content, :board_id, :locked, :sticky)
+    else
+      params.fetch(:message, {}).permit(:subject, :content, :board_id)
+    end
+  end
+
+  def attachments
+    params.permit(attachments: [:file, :description])['attachments']
+  end
+
+  def enumerations
+    acceptable_params = [:active, :is_default, :move_to, :name, :reassign_to_i, :parent_id,
+                         :custom_field_values, :reassign_to_id]
+
+    whitelist = ActionController::Parameters.new
+
+    # Sometimes we receive one enumeration, sometimes many in params, hence
+    # the following branching.
+    if params[:enumerations].present?
+      params[:enumerations].each do |enum, value|
+        enum.tap do
+          whitelist[enum] = {}
+          acceptable_params.each do |param|
+            # We rely on enum being an integer, an id that is. This will blow up
+            # otherwise, which is fine.
+            next if params[:enumerations][enum][param].nil?
+            whitelist[enum][param] = params[:enumerations][enum][param]
+          end
+        end
+      end
+    else
+      params[:enumeration].each do |key, _value|
+        whitelist[key] = params[:enumeration][key]
+      end
+    end
+
+    whitelist.permit!
+  end
+
+  def watcher
+    params.require(:watcher).permit(:watchable, :user, :user_id)
+  end
+
+  def reply
+    params.require(:reply).permit(:content, :subject)
+  end
+
+  def wiki
+    params.require(:wiki).permit(:start_page)
+  end
+
+  def reporting
+    params.fetch(:reporting, {}).permit(:reporting_to_project_id, :reported_project_status_id, :reported_project_status_comment)
+  end
+
+  def membership
+    params.require(:membership).permit(*self.class.permitted_attributes[:membership])
+  end
+
```

Comment:
I think they are very fast to write and execute and prevent unwanted change.

I personally dislike the idea of moving those to the controllers as it will just bloat the controllers and will make reuse quite hard (e.g. between a standard controller and an API one). I'd very much like to sit down with @basiszwo, you and whoever likes to (or is not fast enough too run away in time) and come up with a basic structure for those concerns. I think we should have some guidelines in place so that we don't have to discuss this on every other PR.